simulator/func_sim/alu.h

```diff
@@ -214,6 +214,37 @@ struct ALU
     template<typename I> static void xori( I* instr)  { instr->v_dst = instr->v_src1 ^ instr->v_imm; }
     template<typename I> static void orn( I* instr)   { instr->v_dst = instr->v_src1 | ~instr->v_src2; }
 
+    // Bit permutation
+    template<typename I> static void grev( I* instr)  {
+        typename I::RegisterUInt x = instr->v_src1;
+        size_t bit_width = bitwidth<typename I::RegisterUInt>;
+
+        if (bit_width == 32) {
+            int shamt = instr->v_src2 & 31;
```

Comment:
Please use `shamt_v_src2` function to get `shamt` automatically.

Reply:
How can I use shamt_v_src2 in this case (with 5 or 6 one bits)?

Reply:
https://github.com/MIPT-ILab/mipt-mips/blob/09cdd66e8c75327ccd424905478f4bde31dea749/simulator/func_sim/alu.h#L26

Follow this:
```c++
static_assert(std::is_same_v<T, typename I::RegisterUInt>);
static_assert(std::is_same_v<T, uint32>);
static_assert(bitwidth<T> == 32);
static_assert(log_bitwidth<T> == 5);
static_assert(bitmask<size_t>(log_bitwidth<T>) == 31);
```

```c++
static_assert(std::is_same_v<T, typename I::RegisterUInt>);
static_assert(std::is_same_v<T, uint64>);
static_assert(bitwidth<T> == 64);
static_assert(log_bitwidth<T> == 64);
static_assert(bitmask<size_t>(log_bitwidth<T>) == 63)
```